libsolutil/LP.cpp

```
@@ -0,0 +1,1613 @@
+			vector<rational> data = *left - *right;
+			data[0] *= -1;
+			Constraint c{move(data), _expr.name == "="};
+			if (!tryAddDirectBounds(c))
+			{
+				size_t index = addConstraint(move(c));
+				m_state.back().clauses.emplace_back(Clause{vector<Literal>{Literal{Literal::Constraint, index}}});
+			}
+		}
+		else if (_expr.arguments.at(0).arguments.empty() && isBooleanVariable(_expr.arguments.at(0).name))
+			addBooleanEquality(*parseLiteral(_expr.arguments.at(0)), _expr.arguments.at(1));
+		else if (_expr.arguments.at(1).arguments.empty() && isBooleanVariable(_expr.arguments.at(1).name))
+			addBooleanEquality(*parseLiteral(_expr.arguments.at(1)), _expr.arguments.at(0));
+		else
+		{
+			Literal const& newBoolean = *parseLiteral(declareInternalBoolean());
+			addBooleanEquality(newBoolean, _expr.arguments.at(0));
+			addBooleanEquality(newBoolean, _expr.arguments.at(1));
+		}
+	}
+	else if (_expr.name == "and")
+	{
+		addAssertion(_expr.arguments.at(0));
+		addAssertion(_expr.arguments.at(1));
+	}
+	else if (_expr.name == "or")
+	{
+		// We could try to parse a full clause here.
+		Literal left = parseLiteralOrReturnEqualBoolean(_expr.arguments.at(0));
+		Literal right = parseLiteralOrReturnEqualBoolean(_expr.arguments.at(1));
+		if (left.kind == Literal::Constraint && right.kind == Literal::Constraint)
+		{
+			// We cannot have more than one constraint per clause.
+			right = *parseLiteral(declareInternalBoolean());
+			addBooleanEquality(right, _expr.arguments.at(1));
+		}
+		m_state.back().clauses.emplace_back(Clause{vector<Literal>{left, right}});
+	}
+	else if (_expr.name == "not")
+	{
+		Literal l = negate(parseLiteralOrReturnEqualBoolean(_expr.arguments.at(0)));
+		m_state.back().clauses.emplace_back(Clause{vector<Literal>{l}});
+	}
+	else if (_expr.name == "<=")
+	{
+		optional<vector<rational>> left = parseLinearSum(_expr.arguments.at(0));
+		optional<vector<rational>> right = parseLinearSum(_expr.arguments.at(1));
+		if (!left || !right)
+			return;
```

Should return `unknown` or throw `nonlinear unsupported` or something